### DISCUSSION TITLE
Need help converting keymap to new customized popupmenu

### DISCUSSION THREAD

**Discussion by spamwax** [neoclide/coc.nvim#3985] [Q&A] [answered]
Before the recent updates, following keymap would let me do:

- If popupmenu shows up while I am typing and haven't explicitly selected anything from it, then hitting `<cr>` whould just do a normal enter and go to next line
- If I have selected an item from the popupmenu then that item would get inserted

This was my mapping:
```viml
  inoremap <silent><expr> <cr> pumvisible() && !coc#rpc#request('hasSelected', []) ? "\<C-y>\<CR>" :
        \ pumvisible() && coc#rpc#request('hasSelected', []) ? "\<C-y>" :
        \ "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"
```

How do I convert that to new API. I have tried following, but it's not working the way I want it:

```viml
  inoremap <silent><expr> <cr> coc#pum#visible() && coc#pum#info()['index'] < 0 ? "\<C-y>\<CR>" :
        \ coc#pum#visible() && coc#pum#info()['index'] != -1 ? "\<C-y>" :
        \ "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"
```

What I want is if popumenu is active and I have not selected anything, then hitting `<cr>` should just act as if there was no popup. This is helpful when I want to finish a line and go to next by hitting `<cr>` without selecting any of the suggestions in the popupmenu. Basically, if nothing is selected and `<cr>` is hit, just cancel the popupmenu and do `<cr>`

Same question for the following:

```viml
  inoremap <silent><expr> <TAB>
        \ pumvisible() && !coc#rpc#request('hasSelected', []) ? "\<C-n>" :
        \ pumvisible() && coc#expandableOrJumpable() ? "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
        \ pumvisible() ? "\<C-n>" :
        \ <SID>check_back_space() ? "\<TAB>" :
        \ coc#refresh()
```

**Comment by chemzqm** [comment, marked as answer]
```vim
inoremap <silent><expr> <cr> coc#pum#visible() && coc#pum#info()['index'] != -1 ? coc#pum#confirm() :
        \ "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"
```

**Reply by spamwax** [reply]
Thanks for the quick reply. But I can't get it behave like before. It seems as soon as the popupmenu shows up, the first item is selected automatically!

This is my `<TAB>` and `<CR>` mappings:

```viml
  inoremap <silent><expr> <TAB>
    \ coc#pum#visible() ? coc#_select_confirm() :
    \ coc#expandableOrJumpable() ?
    \ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
    \ <SID>check_back_space() ? "\<TAB>" :
    \ coc#refresh()

  function! s:check_back_space() abort
    let col = col('.') - 1
    return !col || getline('.')[col - 1]  =~# '\s'
  endfunction

  inoremap <silent><expr> <TAB>
    \ coc#pum#visible() ? coc#_select_confirm() :
    \ coc#expandableOrJumpable() ?
    \ "\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\<CR>" :
    \ <SID>check_back_space() ? "\<TAB>" :
    \ coc#refresh()
```

**Reply by chemzqm** [reply]
You have to use `"suggest.noselect": false` to disable automatic selection

**Reply by spamwax** [reply]
Thanks, it should be `"suggest.noselect": true` :)

**Comment by spamwax** [comment]
Is `coc#pum#next(1)` supposed to insert the highlighted completion item from the popup menu at the cursor?
Currently, when I am using the `<TAB>` mapping from `:h coc-completion-example`, and go through each item in the menu, it only inserts completion items that are just one word and does not work with items that are multi words with space in them!
Is this the intended behavior?

https://user-images.githubusercontent.com/1251233/182580853-0c54d26b-556a-4067-8968-1a19ca2bc2bf.mov

```viml
    inoremap <silent><expr> <TAB>
        \ coc#pum#visible() ? coc#pum#next(1):
        \ <SID>check_back_space() ? "\<Tab>" :
        \ coc#refresh()
```

**Reply by chemzqm** [reply]
You have to confirm completion or use configuration `"suggest.invalidInsertCharacters"` to adjust invalid insert characters.

**Comment by lemerchand** [comment]
This doesn't seem to be working for me. The code you suggested only works if I change the keybinding to `<C-CR>`. `<CR>` *is* bound, but only to peartree. Even if I disable it it does not seem to work. Any ideas?

**Reply by arthur-zhuk** [reply]
Looking for a solution as well.

**Reply by chemzqm** [reply]
Your `<CR>` could be remapped by other vim plugin, check your key-mapping by `:verbose imap <CR>`

**Reply by lemerchand** [reply]
I checked `:verbose imap <CR>` before I posted. Like I mentioned, Pairtree is associated with it, but a) shouldn't matter, and b) it still doesn't work after I disable it. :thinking:

**Reply by chemzqm** [reply]
Try minimal vimrc

**Comment by lemerchand** [comment]
Okay. Just tried it with this .vimrc:

```
set nocompatible
filetype plugin indent on
syntax on
set hidden

call plug#begin('~/.vim/autoload/plugged')
Plug 'neoclide/coc.nvim', {'branch': 'release'}
call plug#end()
```

It still did not work. I ran `verbose imap <CR>` just in case. It said 'No mapping found'

**Reply by chemzqm** [reply]
It's expected, you have to remap `<CR>` yourself.

**Reply by lemerchand** [reply]
Omfg duh! Okay, let me give that a try

**Reply by lemerchand** [reply]
Okay, so yea I guess I was wrong. Looks like PearTree was the culprit. The reason it didn't matter even after I disabled it is that I had session options restoring the mappings from a file. When I deleted that along with PearTree `<cr>` began working.

I know it's not your area, but do you have any suggestions for maintaining the functionality of PearTree without the conflict?

**Reply by chemzqm** [reply]
Checkout the doc of PearTree for options to disable the key-mappings created by it and create key-mapping by yourself with it's api.

**Comment by fabOnReact** [comment]
the example still uses tab https://github.com/neoclide/coc-snippets#examples

**Reply by fabOnReact** [reply]
I ended up restoring my previous version using time machine, as currently, I don't have any free time to update my configs.

**Reply by fabOnReact** [reply]
I appreciate the amazing open-source contributions. Sorry, but today is a busy day for me, and can not do this type of upgrade, but hopefully, when I'll be in the holiday I'll do them. 🙏
